File: tests/test_builder.cpp
#include <gtest/gtest.h>

#include "simple_pipe/builder/pipeline_builder.hpp"
#include "simple_pipe/spec/graph_spec.hpp"
#include "test_helper.hpp"

using namespace simple_pipe;
using namespace simple_pipe::test;

TEST(BuilderTest, LoadLinearGraphAndValidate) {
  auto graph = LoadPipelineGraphSpecFromJson(GraphPath("linear"));
  auto flow = BuildFlowChartPipelineSpecFromGraph(graph);
  auto result = ValidateFlowChartPipelineSpec(flow);
  ASSERT_TRUE(result.IsSuccess());
  EXPECT_EQ(flow.nodes.size(), 3u);
}

TEST(BuilderTest, RejectsMultipleSources) {
  auto graph = LoadPipelineGraphSpecFromJson(GraphPath("linear"));
  graph.nodes.push_back(GraphNodeSpec{"app_src2", "app_src", nlohmann::json::object()});
  auto flow = BuildFlowChartPipelineSpecFromGraph(graph);
  auto result = ValidateFlowChartPipelineSpec(flow);
  ASSERT_FALSE(result.IsSuccess());
}

TEST(BuilderTest, InstantiateLinearPipeline) {
  auto graph = LoadPipelineGraphSpecFromJson(GraphPath("linear"));
  auto flow = BuildFlowChartPipelineSpecFromGraph(graph);
  ASSERT_TRUE(ValidateFlowChartPipelineSpec(flow).IsSuccess());
  auto built = BuildFlowChartPipelineFromSpec(flow);
  EXPECT_EQ(built.nodes.size(), 3u);
  EXPECT_NE(built.source, nullptr);
  EXPECT_NE(built.sink, nullptr);
}

File: tests/test_builder.py
from pathlib import Path

import pytest

from simple_pipe.builder.pipeline_builder import (
    build_flow_chart_pipeline_spec_from_graph,
    validate_flow_chart_pipeline_spec,
)
from simple_pipe.spec.graph_spec import load_graph_spec_from_json

ROOT = Path(__file__).resolve().parents[1]


def test_load_linear_graph():
    graph = load_graph_spec_from_json(ROOT / "configs/demo/graph/linear.json")
    flow = build_flow_chart_pipeline_spec_from_graph(graph)
    result = validate_flow_chart_pipeline_spec(flow)
    assert result.ok
    assert len(flow.nodes) == 3


def test_validate_rejects_duplicate_src():
    graph = load_graph_spec_from_json(ROOT / "configs/demo/graph/linear.json")
    graph.nodes.append(graph.nodes[0].__class__(id="app_src2", type="app_src"))
    flow = build_flow_chart_pipeline_spec_from_graph(graph)
    result = validate_flow_chart_pipeline_spec(flow)
    assert not result.ok
